contracts/facets/ThirdPartyFacet.sol

```diff
@@ -91,117 +89,124 @@ contract ThirdPartyFacet {
-                // then don't add to unaccusedArchsAddresses
-                if (bondedArchaeologists[i] == accusedArchAddresses[j]) {
-                    isUnaccused = false;
-                    break;
+        // if none of the accusals were valid because the archaeologists have already been accused, return without taking action
+        if (accusalCount == 0) {
+            return;
+        }
+
+
+        address[] memory archaeologistAddresses = s.sarcophagi[sarcoId].archaeologists;
+
+        // the sarcophagus is compromised if k or more archaeologists have been accused successfully on this call
+        bool isSarcophagusCompromised = accusalCount >= sarco.minShards;
+
+        // if the current call hasn't resulted in at least k archaeologists being accused
+        // check if total number of historical accusals on sarcophagus is greater than k
+        uint historicalAccusals = 0;
+        if (!isSarcophagusCompromised) {
+            for (uint256 i = 0; i < archaeologistAddresses.length; i++) {
+                if (s.sarcophagusArchaeologists[sarcoId][archaeologistAddresses[i]].accused) {
+                    historicalAccusals++;
                 }
             }
+            // the sarcophagus is compromised if k or more archaeologists have been accused over the lifetime of the sarcophagus
+            if (historicalAccusals >= sarco.minShards) {
+                isSarcophagusCompromised = true;
+            }
+        }
 
-            // If this arch address wasn't in the accused list, free it from its curse
-            if (isUnaccused) {
-                // There are technically no rewards here, since the sarcophagus
-                // has been compromised, so here this effectively merely resets
-                // the state of the non-malicious archaeologists, as if they never
-                // bonded to this sarcophagus in the first place.
-                //
-                // Of course, whatever rewards they might have gained in previous
-                // rewraps remains theirs.
-                LibBonds.freeArchaeologist(sarcoId, bondedArchaeologists[i]);
+        // if k or more archaeologists have been accused over the lifetime of the sarcophagus, funds should
+        // be returned to the remaining well behaved archaeologists
+        if (isSarcophagusCompromised) {
+            // update the sarcophagus state to Accused, the outer key is compromised
+            sarco.state = LibTypes.SarcophagusState.Accused;
+
+            // iterate through all archaeologist addresses on the sarcophagus
+            for (uint256 i = 0; i < archaeologistAddresses.length; i++) {
+                // if the archaeologist has never been accused, release their locked bond back to them
+                if (!s.sarcophagusArchaeologists[sarcoId][archaeologistAddresses[i]].accused) {
+                     LibBonds.freeArchaeologist(sarcoId, archaeologistAddresses[i]);
```

Comment:
In the original version of this function, we don't refund the digging fees allocated to well-behaved archaeologists back to the embalmer. In this PR I've preserved that functionality but I think we ought to provide the full refund because the archaeologists aren't receiving those digging fees if the sarcophagus is being closed down and the embalmer isn't receiving the services they've paid for.

Reply:
Agreed

Reply:
The first loop only runs through archaeologists that have leaked keyshares and not the full list of archaeologists on the sarcophagus, so unfortunately there's no way to track historical accusations at that stage. Good thought though, I optimized as much as possible but the looping in this function is pretty aggressive anyway.